urllib3/connection.py

```diff
@@ -162,6 +162,34 @@ def connect(self):
         conn = self._new_conn()
         self._prepare_conn(conn)
 
+    def request_chunked(self, method, url, body=None, headers={}):
+        """
+        Alternative to the common request method, which sends the
+        body with chunked encoding and not as one block
+        """
+        header_names = set(k.lower() for k in headers)
+        self.putrequest(method, url, skip_accept_encoding=True)
+        for header, value in headers.items():
+            self.putheader(header, value)
+        if 'transfer-encoding' not in header_names:
+            self.putheader('Transfer-Encoding', 'chunked')
+        self.endheaders()
+
+        if body is not None:
+            if isinstance(body, six.string_types):
+                body = [body]
+            for chunk in body:
+                if not chunk:
+                    continue
+                len_str = hex(len(chunk))[2:]
+                self.send(len_str.encode('utf-8'))
+                self.send(b'\r\n')
+                self.send(chunk.encode('utf-8'))
+                self.send(b'\r\n')
```

Comment:
Python 2.7's httplib is not the bar of acceptable that we're trying to meet, Python 3's is. The odds of a Python 2 user passing a unicode chunk here is low, but if they do it we should either error helpfully (rather than letting the implicit encode/decode catch it) or we should try to help them (I'd prefer to do the first of those things). However, the odds of a Python 3 user passing bytes here is very high, and we absolutely **must** do the right thing here in the face of that situation.

To be clear, here is the space of things a user might do and whether this function works correctly:

|                            | Python 2                           | Python 3                           |
|----------------------------|------------------------------------------|----------------------------------|
| Bytestring: only ASCII     | Works (only on some platforms, may fail) | **Fails**: `AttributeError`        |
| Bytestring: not ASCII      | **Fails**: `UnicodeDecodeError`    | **Fails**: `AttributeError`          |
| Unicode string: only ASCII | Works                              | Works                              |
| Unicode string: not ASCII  | **Invalid Output**: Incorrect chunk header. | **InvalidOutput**: Incorrect chunk header. |

It is not acceptable to fail in the face of a bytestring, so all of those failures need to be dealt with. It's also unacceptable to generate invalid output, because this amounts to silent data corruption, which we must not do.

I strongly recommend adding tests for each of the four data types I've listed there and confirm that they all work.